### PR TITLE
added feature flags

### PR DESCRIPTION
**Pull request by komakurae**
to check how it works go to
auth_bloc.dart

there you  can change user id - 1 (feature disabled); id - 2 (feature enabled)

### CODE REVIEW

File: lib/blocs/feature_flag/feature_flag_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  final FeatureFlagClient featureFlagClient;
+
+  late StreamSubscription<Flag> _subscription;
+
+  FeatureFlagBloc({
+    required this.featureFlagClient,
+  }) : super(const FeatureFlagState()) {
+    on<_Initialize>(_onInitFlag);
+    on<_Fetch>(_onFetchFlag);
+    on<_Reload>(_onReloadFlag);
```
Comment (StepanchukPv) [inline lib/blocs/feature_flag/feature_flag_bloc.dart]: You can replace _Fetch and _Reload with _Load event

File: lib/blocs/feature_flag/feature_flag_state.dart [hunk 2]
```diff
@@ -0,0 +1,27 @@
+part of 'feature_flag_bloc.dart';
+
+@freezed
+class FeatureFlagState with _$FeatureFlagState {
+  const FeatureFlagState._();
+
+  const factory FeatureFlagState({
+    @Default([]) List<Flag> flags,
```
Comment (StepanchukPv) [inline lib/blocs/feature_flag/feature_flag_state.dart]: 1. Flags should have default value 
2. It's better to store flags inside Map

File: lib/blocs/feature_flag/feature_flag_state.dart [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    @Default(FeatureFlagStatus.loading) FeatureFlagStatus status,
+    Identity? identity,
+  }) = _FeatureFlagState;
+
+  bool isEnabled(String flagName) {
+    final found = flags.firstWhereOrDefault(
+      (value) =>
+          value.feature.name == flagName &&
+          ((value.enabled != null) && value.enabled! == true),
```
Comment (StepanchukPv) [inline lib/blocs/feature_flag/feature_flag_state.dart]: reduntant

File: lib/screens/home/home_screen.dart [hunk 4]
```diff
@@ -12,43 +15,49 @@ class HomeScreen extends StatelessWidget {
...
-      ],
-      bottomNavigationBuilder: (_, tabsRouter) {
-        return DecoratedBox(
-          decoration: const BoxDecoration(
-            boxShadow: [
-              BoxShadow(
-                spreadRadius: 1,
-                blurRadius: 3,
-                offset: Offset(0, 3),
+    return BlocBuilder<FeatureFlagBloc, FeatureFlagState>(
```
Comment (StepanchukPv) [inline lib/screens/home/home_screen.dart]: It's better to user BlocSelector or context.select to track only feature_flag changes

File: lib/screens/home/home_state_wrapper.dart [hunk 5]
```diff
@@ -12,8 +13,20 @@ class HomeStateWrapper extends StatelessWidget {
 
   @override
   Widget build(BuildContext context) {
+    final userEmail = context.read<AuthBloc>().state.userProfile.email;
     return MultiBlocProvider(
       providers: [
+        BlocProvider(
+          create: (context) => getIt<FeatureFlagBloc>()
```
Comment (StepanchukPv) [inline lib/screens/home/home_state_wrapper.dart]: You can pass FeatureFlags bloc to AuthBloc to await fetch feature flags

File: lib/services/feature_flag/feature.dart [hunk 6]
```diff
@@ -0,0 +1,9 @@
+enum Feature {
+  cartsTab,
+}
+
+extension FeatureX on Feature {
+  String get id => switch (this) {
```
Comment (StepanchukPv) [inline lib/services/feature_flag/feature.dart]: You can declare id inside Feature -> Feature(this.id) and default value

File: lib/services/feature_flag/feature_flag_client.dart [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import 'package:shop_app/core/constants/env_constants.dart';
+
+class FeatureFlagClient {
+  static late FlagsmithClient _flagsmith;
+
+  static Future<FeatureFlagClient> init() async {
+    _flagsmith = await FlagsmithClient.init(
+      apiKey: EnvConstants.appFlagsmithKey,
+    );
+    await _flagsmith.getFeatureFlags();
```
Comment (StepanchukPv) [inline lib/services/feature_flag/feature_flag_client.dart]: init should not be responsible for loading data

**Review by StepanchukPv — CHANGES_REQUESTED** [review]

File: lib/screens/home/home_state_wrapper.dart [hunk 8]
```diff
@@ -13,19 +12,10 @@ class HomeStateWrapper extends StatelessWidget {
...
-          create: (context) => getIt<FeatureFlagBloc>()
-            ..add(
-              FeatureFlagEvent.initialize(
-                featureName: Feature.cartsTab.id,
-                identity: fs.Identity(
-                  identifier: userEmail,
-                ),
-              ),
-            ),
+          create: (context) => getIt<FeatureFlagBloc>(),
```
Comment (StepanchukPv) [inline lib/screens/home/home_state_wrapper.dart]: Move to AppStateWrapper

File: lib/blocs/auth/auth_bloc.dart [hunk 9]
```diff
@@ -39,9 +45,16 @@ class AuthBloc extends Bloc<AuthEvent, AuthState> {
...
-          2,
+          1,
         ); // for id = 1 feature disabled, for id = 2 enabled
 
+        featureFlagBloc.add(
+          FeatureFlagEvent.initialize(
+            featureName: Feature.cartsTab.featureName,
+            identity: fs.Identity(identifier: userProfile.email),
+          ),
+        );
```
Comment (StepanchukPv) [inline lib/blocs/auth/auth_bloc.dart:56]: You need to await for featureFlag to load. FeatureFlagBloc.stream.firstWhere((state) => state.status == FeatureFlagStatus.complete)

File: lib/blocs/feature_flag/feature_flag_bloc.dart [hunk 10]
```diff
@@ -22,9 +21,9 @@ class FeatureFlagBloc extends Bloc<FeatureFlagEvent, FeatureFlagState> {
     required this.featureFlagClient,
   }) : super(const FeatureFlagState()) {
     on<_Initialize>(_onInitFlag);
-    on<_Fetch>(_onFetchFlag);
-    on<_Reload>(_onReloadFlag);
+    on<_Load>(_onLoadFlags);
     on<_Register>(_onRegisterFlag);
+    on<_ChangeIdentity>(_changeIdentityFlag);
```
Comment (StepanchukPv) [inline lib/blocs/feature_flag/feature_flag_bloc.dart:27]: Do you use this event?

**Review by StepanchukPv — APPROVED** [review]

File: lib/blocs/feature_flag/feature_flag_bloc.dart (hunk 10 above)
Comment (komakurae) [reply lib/blocs/feature_flag/feature_flag_bloc.dart:27]: not for now, but maybe later I will add change user method, without complete log out

Comment (komakurae) [reply lib/blocs/feature_flag/feature_flag_bloc.dart:27]: that's why I may need it, but not sure for now

**Pull request merged**